drivers/blk/mmc/imx/usdhc.c

```diff
@@ -0,0 +1,1030 @@
+
+            status = SUCCESS;
+            success_count = 1;
+            break;
+
+        case WRITE_BLOCKS:
+            poll = usdhc_write_blocks(req_offset, req_block_number * block_to_sectors,
+                                      req_count * block_to_sectors);
+            if (poll == PollPending) {
+                return;
+            }
+            driver_state.data_transfer = DataStateInit;
+
+            status = SUCCESS;
+            success_count = 1;
+            break;
+
+        case FLUSH:
+        case BARRIER:
+            /* No-ops. */
+            status = SUCCESS;
+            success_count = req_count;
+            break;
+
+        default:
+            LOG_DRIVER_ERR("Unknown command code: %d\n", req_code);
+            return;
+        }
+
+        int err = blk_enqueue_resp(&blk_queue, status, success_count, req_id);
+        assert(!err);
+        LOG_DRIVER("Enqueued response: status=%d, success_count=%d, id=%d\n", status, success_count, req_id);
+        microkit_notify(USDHC_VIRT_CHANNEL);
+
+        driver_state.clients = ClientStateIdle;
+        return handle_clients();
+    }
+
+    default:
+        assert(!"unreachable");
+    }
+}
+
+void usdhc_executor(bool is_irq)
+{
+    if (driver_state.executor == ExecutorStateInit && !is_irq) {
+        /* Ignore client requests until we're ready (at which point we will process them) */
+        return;
+    }
+
+    poll_t poll;
+    switch (driver_state.executor) {
+    case ExecutorStateInit:
+        poll = perform_card_identification_and_select();
+        if (poll == PollPending) {
+            return;
+        }
+        setup_blk_queues();
+
+        driver_state.executor = ExecutorStateActive;
+        fallthrough;
+
+    case ExecutorStateActive:
+        handle_clients();
+        // We always stay in the Active state now.
+        break;
+
+    default:
+        assert(!"unreachable");
+    }
+}
+
+void notified(microkit_channel ch)
+{
+    switch (ch) {
+    case USDHC_IRQ_CHANNEL:
+        usdhc_executor(true);
+        break;
+
+    case USDHC_VIRT_CHANNEL:
+        usdhc_executor(false);
+        break;
+
+    case USDHC_TIMER_CHANNEL:
+        LOG_DRIVER("got timer interrupt -- UNHANDLED\n");
+        assert(false);
+        break;
+
+    default:
+        LOG_DRIVER_ERR("notification on unknown channel: %d\n", ch);
+        break;
+    }
+
+    if (ch == USDHC_IRQ_CHANNEL) {
+        microkit_irq_ack(ch);
+    }
+}
+
+void init()
+{
+    reset_driver_and_card_state();
+
+    LOG_DRIVER("Beginning driver initialisation...\n");
+
+    usdhc_reset();
+    usdhc_executor(true);
+
+    // TODO: This appears to be broken and does not work at all; Linux does not
+    //       even notice when the card is inserted/removed....
+    // if (usdhc_regs->pres_state & USDHC_PRES_STATE_CINST) {
+    //     sddf_printf("card inserted\n");
+    // } else {
+    //     sddf_printf("card not inserted or power on reset\n");
+    // }
```

Comment:
Are we convinced that card detect is broken? If we are then I guess we should maybe remove this TODO?

This reminds me, maybe we should mention that card detect does not work in the README for the driver?